### PR TITLE
un-implement Request from *proto.BatchRequest

### PR DESCRIPTION
**Pull request by tbg**
and various other progress towards #2143.

### CODE REVIEW

**Comment by tschottdorf** [comment]
@spencerkimball your eyes on the `add PushNow to PushRequest` commit are appreciated. I think that allows us to completely get rid of explicit request timestamps (except when carried in a `Txn`).

**Comment by tschottdorf** [comment]
thanks @tamird, feedback is addressed (fixed most of it in the original commits, which unfortunately resulted in your comments being removed from the thread here).

File: client/batch.go [hunk 1]
```diff
@@ -194,20 +193,22 @@ func (b *Batch) fillResults(br *proto.BatchResponse, pErr *proto.Error) error {
...
-func (b *Batch) InternalAddRequest(args proto.Request) {
-	numRows := 0
-	switch args.(type) {
-	case *proto.GetRequest,
-		*proto.PutRequest,
-		*proto.ConditionalPutRequest,
-		*proto.IncrementRequest,
-		*proto.DeleteRequest:
-		numRows = 1
+// InternalAddRequest adds the specified request to the batch. It is intended
```
Comment (tamird) [inline client/batch.go]: s/request/requests/

File: kv/dist_sender.go [hunk 2]
```diff
@@ -444,7 +441,7 @@ func (ds *DistSender) sendAttempt(trace *tracer.Trace, ba proto.BatchRequest, de
 	}
 
 	// TODO(tschottdorf) &ba -> ba
```
Comment (tamird) [inline kv/dist_sender.go]: this is still here

**Comment by tamird** [comment]
LGTM but wait for @spencerkimball

File: kv/dist_sender.go (hunk 2 above)
Comment (tschottdorf) [inline kv/dist_sender.go]: Whoops. Definitely got it now.

File: client/batch.go (hunk 1 above)
Comment (tschottdorf) [inline client/batch.go]: Done.

File: kv/txn_coord_sender.go [hunk 3]
```diff
@@ -322,8 +322,7 @@ func (tc *TxnCoordSender) Send(ctx context.Context, ba proto.BatchRequest) (*pro
 	// TxnID (if applicable). Begin a Trace which follows this request.
 	trace := tc.tracer.NewTrace(&ba)
 	defer trace.Finalize()
-	// TODO(tschottdorf): always "Batch"
-	defer trace.Epoch(fmt.Sprintf("sending %s", ba.Method()))()
+	defer trace.Epoch(fmt.Sprintf("sending %s", proto.Batch))()
```
Comment (bdarnell) [inline kv/txn_coord_sender.go]: I don't really like this change from `.Method()` to a hard-coded `proto.Batch` (here and elsewhere, especially where it's combined with a literal like in dist_sender_test.go). I'd prefer either keeping a `.Method()` method on `BatchRequest` (similar to the one we have on `sql/driver.Request`, even though the only method there is `Execute`), or dispense with the constant altogether and just make it a string literal: `trace.Epoch("sending Batch")`.

**Comment by bdarnell** [comment]
LGTM

File: kv/txn_coord_sender.go (hunk 3 above)
Comment (tschottdorf) [inline kv/txn_coord_sender.go]: I've replaced uses of `proto.Batch` almost everywhere. There remain, I think, two places where a TODO is in place to get rid of them already.

**Comment by tschottdorf** [comment]
I added two more commits to deal with flaky `TestStoreRangeRebalance`: one that turns of txn auto-gc and the other that fixes up some code in the "multiplexing" test `rpcSender` we use for `multiStoreSender` (so that it bails if it hits the right range and gets an error; previously it'd just try the next which could lead to infinite retries if your last store wasn't the right one and would give you key mismatch errors)

**Comment by tamird** [comment]
still LGTM

**Comment by bdarnell** [comment]
LGTM

**Comment by mrtracy** [comment]
Also LGTM

**Pull request merged**